xds/internal/balancer/gracefulswitch/gracefulswitch.go

```diff
@@ -0,0 +1,368 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package gracefulswitch implements a graceful switch load balancer.
+package gracefulswitch
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"google.golang.org/grpc/balancer"
+	"google.golang.org/grpc/balancer/base"
+	"google.golang.org/grpc/connectivity"
+	"google.golang.org/grpc/resolver"
+)
+
+var (
+	errBalancerClosed = errors.New("gracefulSwitchBalancer is closed")
+)
+
+// NewBalancer returns a graceful switch Balancer.
+func NewBalancer(cc balancer.ClientConn, opts balancer.BuildOptions) *Balancer {
+	return &Balancer{
+		cc:    cc,
+		bOpts: opts,
+	}
+}
+
+// Balancer is a utility to gracefully switch from one balancer to
+// a new balancer.
+type Balancer struct {
+	bOpts balancer.BuildOptions
+	cc    balancer.ClientConn
+
+	// mu protects the following fields and all fields within balancerCurrent
+	// and balancerPending. mu does not need to be held when calling into the
+	// child balancers, as all calls into these children happen only as a direct
+	// result of a call into the gracefulSwitchBalancer, which are also
+	// guaranteed to be synchronous. There is one exception: an UpdateState call
+	// from a child balancer when current and pending are populated can lead to
+	// calling Close() on the current. To prevent that racing with an
+	// UpdateSubConnState from the channel, we hold currentMu during Close and
+	// UpdateSubConnState calls.
+	mu              sync.Mutex
+	balancerCurrent *balancerWrapper
+	balancerPending *balancerWrapper
+	closed          bool // set to true when this balancer is closed
+
+	// currentMu must be locked before mu. This mutex guards against this
+	// sequence of events: UpdateSubConnState() called, finds the
+	// balancerCurrent, gives up lock, updateState comes in, causes Close() on
+	// balancerCurrent before the UpdateSubConnState is called on the
+	// balancerCurrent.
+	currentMu sync.Mutex
+}
+
+// swap swaps out the current lb with the pending LB and updates the ClientConn.
+// The caller must hold gsb.mu.
+func (gsb *Balancer) swap() {
+	gsb.cc.UpdateState(gsb.balancerPending.lastState)
+	cur := gsb.balancerCurrent
+	gsb.balancerCurrent = gsb.balancerPending
+	gsb.balancerPending = nil
+	go func() {
+		gsb.currentMu.Lock()
+		defer gsb.currentMu.Unlock()
+		cur.Close()
+	}()
+}
+
+// Helper function that checks if the balancer passed in is current or pending.
+// The caller must hold gsb.mu.
+func (gsb *Balancer) balancerCurrentOrPending(bw *balancerWrapper) bool {
+	return bw == gsb.balancerCurrent || bw == gsb.balancerPending
+}
+
+// SwitchTo gracefully switches to the new balancer. This function must be
+// called synchronously alongside the rest of the balancer.Balancer methods this
```

Comment:
> This function must be called synchronously alongside the rest of the balancer.Balancer methods

Does this mean that we are explicitly disallowing the graceful switch balancer to be the top-level balancer on a channel? We don't have a `balancer.Builder` for this balancer, but grpc could still use this as a top-level balancer if it wanted to. 

The reason I'm is because the [ccBalancerWrapper](https://github.com/grpc/grpc-go/blob/master/balancer_conn_wrappers.go#L46) synchronizes some methods on the balancer (for example, `UpdateSubConnState` and `ExitIdle`) by having them go through a channel which is read by a watcher, but it directly forwards some other methods (for example, `UpdateClientConnState` and `ResolverError`). In this case if grpc were to use this graceful switch balancer as the top-level balancer, how would it guarantee that calls to `SwitchTo` are synchronized with other calls to the `balancer.Balancer` methods?

Reply:
I re-read the `balancer_conn_wrapper` code and looks like all calls to the wrapped balancer is guarded by the `balancerMu` even though some are handled inline, while some are pushed on a channel and handled in the `watcher` goroutine. So, this should be fine as it is. Thanks.